tinylfu/tinylfu.go

```diff
@@ -0,0 +1,122 @@
+// Package tinylfu is an implementation of the W-TinyLFU caching algorithm.
+// See details at http://arxiv.org/abs/1512.00727
+package tinylfu
+
+// Policy implements a windowed TinyLFU eviction policy. It is not safe for concurrent access.
+type Policy struct {
+	data     map[uint64]*element
+	admittor AdmissionPolicy
+	stats    StatsRecorder
+
+	window    *list
+	probation *list
+	protected *list
+
+	capacity     int
+	maxWindow    int
+	maxProtected int
+}
+
+// New creates a new TinyLFU cache.
+func New(capacity int, opts ...Option) *Policy {
+	// Consistent behavior relies on capacity for one element in each segment.
+	if capacity < 3 {
+		panic("tinylfu: capacity must be positive")
+	}
+
+	p := &Policy{
+		data:      make(map[uint64]*element),
+		window:    newList(),
+		probation: newList(),
+		protected: newList(),
+		capacity:  capacity,
+	}
+
+	WithSegmentation(0.99, 0.8)(p)
```

Comment:
Since the ratios might depend on the workloads, why not take these values through Option and fallback to these if not given ?